rust-engine/src/backends/lean.rs

```diff
@@ -154,568 +66,364 @@ impl Lean {
-                        .append(allocator.softline())
-                        .append(allocator.intersperse(args, allocator.softline()))
-                        .parens()
-                        .group()
-                }
+        fn pat_kind(&'a self, pat_kind: &'b PatKind) -> DocBuilder<'a, Self, A> {
+            match pat_kind {
+                PatKind::Wild => docs!["_"],
+                PatKind::Ascription { pat, ty: _ } => docs![pat],
+                PatKind::Binding {
+                    mutable,
+                    var,
+                    mode,
+                    sub_pat,
+                } => match (mutable, mode, sub_pat) {
+                    (false, BindingMode::ByValue, None) => docs![var],
+                    _ => panic!(),
+                },
+                _ => todo!(),
             }
-            TyKind::Arrow { inputs, output } => docs![
-                allocator,
-                allocator.concat(inputs.into_iter().map(|input| docs![
-                    allocator,
-                    input,
-                    allocator.reflow(" -> ")
-                ])),
-                "Result ",
-                output
-            ]
-            .parens(),
-            TyKind::Ref {
-                inner: _,
-                mutable: _,
-                region: _,
-            } => print_todo!(allocator),
-            TyKind::Param(local_id) => local_id.pretty(allocator),
-            TyKind::Slice(ty) => docs![allocator, "Array ", ty].parens(),
-            TyKind::Array { ty, length } => {
-                docs![allocator, "Vector ", ty, allocator.softline(), &(**length),]
-                    .parens()
-                    .group()
-            }
-            TyKind::RawPointer => print_todo!(allocator),
-            TyKind::AssociatedType { impl_: _, item: _ } => print_todo!(allocator),
-            TyKind::Opaque(_global_id) => print_todo!(allocator),
-            TyKind::Dyn(_dyn_trait_goals) => print_todo!(allocator),
-            TyKind::Resugared(_resugared_ty_kind) => print_todo!(allocator),
-            TyKind::Error(_diagnostic) => print_todo!(allocator),
         }
-    }
-}
-
-impl<'a, 'b> Pretty<'a, Allocator<Lean>, Span> for &'b SpannedTy {
-    fn pretty(self, allocator: &'a Allocator<Lean>) -> DocBuilder<'a, Allocator<Lean>, Span> {
-        self.ty.pretty(allocator)
-    }
-}
 
-impl<'a, 'b> Pretty<'a, Allocator<Lean>, Span> for &'b PrimitiveTy {
-    fn pretty(self, allocator: &'a Allocator<Lean>) -> DocBuilder<'a, Allocator<Lean>, Span> {
-        match self {
-            PrimitiveTy::Bool => allocator.text("Bool"),
-            PrimitiveTy::Int(int_kind) => int_kind.pretty(allocator),
-            PrimitiveTy::Float(float_kind) => float_kind.pretty(allocator),
-            PrimitiveTy::Char => allocator.text("Char"),
-            PrimitiveTy::Str => allocator.text("String"),
+        fn ty(&'a self, ty: &'b Ty) -> DocBuilder<'a, Self, A> {
+            ty.kind().pretty(self)
```

Comment:
WDYT about `docs![ty.kind()]` here instead?
It's a matter of taste, I'm thinking maybe it is good for uniformity?